File: tests/test_adapters.py
from __future__ import annotations

import unittest
from pathlib import Path
from tempfile import NamedTemporaryFile

from tui.adapters import (
    BasicHTTPAgentAdapter,
    CodexAdapter,
    CustomPythonAdapter,
    HermesAdapter,
    OpenCLIAdapter,
    OpenClawAdapter,
    PiAdapter,
    adapter_for,
    detect_available_adapters,
)
from tui.agent_runtime import AgentContext
from tui.agent_runtime import AgentContext


class AdapterFactoryTest(unittest.TestCase):
    def test_hermes_identifier_returns_hermes_adapter(self):
        adapter = adapter_for("hermes", timeout_sec=120)
        self.assertIsInstance(adapter, HermesAdapter)

    def test_openclaw_identifier_returns_openclaw_adapter(self):
        adapter = adapter_for("openclaw")
        self.assertIsInstance(adapter, OpenClawAdapter)

    def test_pi_identifier_returns_pi_adapter(self):
        adapter = adapter_for("pi")
        self.assertIsInstance(adapter, PiAdapter)

    def test_custom_python_identifier_returns_custom_python_adapter(self):
        adapter = adapter_for("custom-python:./agent.py")
        self.assertIsInstance(adapter, CustomPythonAdapter)
        self.assertIn("./agent.py", adapter.command_template)

    def test_basic_identifier_returns_basic_adapter(self):
        adapter = adapter_for("basic")
        self.assertIsInstance(adapter, BasicHTTPAgentAdapter)
        self.assertIn("basic_http_agent.py", adapter.command_template[1])

    def test_opencli_identifier_returns_opencli_adapter(self):
        adapter = adapter_for("opencli")
        self.assertIsInstance(adapter, OpenCLIAdapter)

    def test_codex_identifier_returns_codex_adapter(self):
        adapter = adapter_for("codex")
        self.assertIsInstance(adapter, CodexAdapter)

    def test_unknown_identifier_falls_back_to_custom_command(self):
        from tui.agent_runtime import CustomCommandAdapter
        adapter = adapter_for("echo FLAG{test}")
        self.assertIsInstance(adapter, CustomCommandAdapter)


class HermesAdapterTest(unittest.TestCase):
    def setUp(self):
        self.ctx = AgentContext(
            match_id="match_001", room_id="room_x", team_id="team_a",
            phase="ATTACK",
            targets=[{"team_id": "team_b", "base_url": "http://127.0.0.1:18082"}],
            local_target={"base_url": "http://127.0.0.1:18081"},
            allowed_targets=["http://127.0.0.1:18081", "http://127.0.0.1:18082"],
            target_template_id="real_ctf_web_awd_01",
        )

    def test_command_template_uses_dash_z_prompt(self):
        adapter = HermesAdapter(hermes_bin="hermes")
        adapter.configure(self.ctx)
        self.assertIn("-z", adapter.command_template)
        self.assertIn("--yolo", adapter.command_template)
        template_str = " ".join(adapter.command_template)
        self.assertIn("{target_url}", template_str)

    def test_model_flag_added(self):
        adapter = HermesAdapter(hermes_bin="hermes", model="claude-opus-4-8")
        self.assertIn("-m", adapter.command_template)
        self.assertIn("claude-opus-4-8", adapter.command_template)

    def test_runs_echo_as_hermes_simulating_flag_find(self):
        adapter = HermesAdapter(hermes_bin="echo")
        adapter.configure(self.ctx)
        result = adapter.run()
        self.assertTrue(result.ok)
        self.assertEqual(len(result.actions), 1)
        self.assertTrue(result.actions[0].ok)


class OpenClawAdapterTest(unittest.TestCase):
    def test_command_template_uses_infer_model_run(self):
        adapter = OpenClawAdapter(openclaw_bin="echo")
        template = adapter.command_template
        self.assertIn("infer", template)
        self.assertIn("model", template)
        self.assertIn("run", template)
        self.assertIn("--local", template)
        self.assertIn("--json", template)

    def test_model_flag_added_when_specified(self):
        adapter = OpenClawAdapter(openclaw_bin="echo", model="claude-sonnet-4-6")
        template = adapter.command_template
        self.assertIn("--model", template)
        self.assertIn("claude-sonnet-4-6", template)


class PiAdapterTest(unittest.TestCase):
    def test_command_template_uses_print_mode_json(self):
        adapter = PiAdapter(pi_bin="echo")
        self.assertIn("--print", adapter.command_template)
        self.assertIn("--mode", adapter.command_template)
        self.assertIn("json", adapter.command_template)
        template_str = " ".join(adapter.command_template)
        self.assertIn("{target_url}", template_str)

    def test_model_flag_added(self):
        adapter = PiAdapter(pi_bin="echo", model="claude-sonnet-4-6")
        self.assertIn("--model", adapter.command_template)
        self.assertIn("claude-sonnet-4-6", adapter.command_template)


class CustomPythonAdapterTest(unittest.TestCase):
    def test_runs_script_with_target_url(self):
        with NamedTemporaryFile(mode="w", suffix=".py", delete=False) as f:
            f.write("import sys; print(f'FLAG{{script_capture}} for {sys.argv[1]}')")
            script_path = f.name

        try:
            ctx = AgentContext(
                match_id="match_001", room_id="room_x", team_id="team_a",
                phase="ATTACK",
                targets=[{"team_id": "team_b", "base_url": "http://127.0.0.1:18082"}],
                local_target={"base_url": "http://127.0.0.1:18081"},
                allowed_targets=["http://127.0.0.1:18081", "http://127.0.0.1:18082"],
                target_template_id="real_ctf_web_awd_01",
            )
            adapter = CustomPythonAdapter(script_path=script_path)
            adapter.configure(ctx)
            result = adapter.run()
            self.assertTrue(result.ok)
            self.assertIn("FLAG{script_capture}", result.flags_captured)
        finally:
            Path(script_path).unlink()


class OpenCLIAdapterTest(unittest.TestCase):
    def test_command_template_uses_browser_extract(self):
        adapter = OpenCLIAdapter(opencli_bin="echo")
        self.assertIn("browser", adapter.command_template)
        self.assertIn("extract", adapter.command_template)

    def test_runs_open_and_extract_via_echo(self):
        ctx = AgentContext(
            match_id="match_001", room_id="room_x", team_id="team_a",
            phase="ATTACK",
            targets=[{"team_id": "team_b", "base_url": "http://127.0.0.1:18082"}],
            local_target={"base_url": "http://127.0.0.1:18081"},
            allowed_targets=["http://127.0.0.1:18081", "http://127.0.0.1:18082"],
            target_template_id="real_ctf_web_awd_01",
        )
        adapter = OpenCLIAdapter(opencli_bin="echo")
        adapter.configure(ctx)
        result = adapter.run()
        self.assertTrue(result.ok)


class CodexAdapterTest(unittest.TestCase):
    def test_command_template_uses_codex_exec(self):
        adapter = CodexAdapter(codex_bin="echo")
        self.assertIn("exec", adapter.command_template)

    def test_prompt_includes_target_placeholder(self):
        adapter = CodexAdapter(codex_bin="echo")
        template_str = " ".join(adapter.command_template)
        self.assertIn("{target_url}", template_str)


class DetectAvailableAdaptersTest(unittest.TestCase):
    def test_returns_dict_with_expected_keys(self):
        available = detect_available_adapters()
        for key in ("hermes", "openclaw", "opencli", "codex", "pi", "python3"):
            self.assertIn(key, available)

File: client/test-adapters.js
"use strict";

const { describe, it } = require("node:test");
const assert = require("node:assert");
const {
  HermesAdapter,
  OpenClawAdapter,
  OpenCLIAdapter,
  CodexAdapter,
  PiAdapter,
  CustomPythonAdapter,
  adapterFor,
  detectAvailableAdapters,
} = require("./adapters.js");
const { CustomCommandAdapter } = require("./agentRuntime.js");

describe("adapterFor", () => {
  it("returns HermesAdapter for hermes identifier", () => {
    const adapter = adapterFor("hermes");
    assert.ok(adapter instanceof HermesAdapter);
  });

  it("returns OpenClawAdapter for openclaw identifier", () => {
    const adapter = adapterFor("openclaw");
    assert.ok(adapter instanceof OpenClawAdapter);
  });

  it("returns PiAdapter for pi identifier", () => {
    const adapter = adapterFor("pi");
    assert.ok(adapter instanceof PiAdapter);
  });

  it("returns OpenCLIAdapter for opencli identifier", () => {
    const adapter = adapterFor("opencli");
    assert.ok(adapter instanceof OpenCLIAdapter);
  });

  it("returns CodexAdapter for codex identifier", () => {
    const adapter = adapterFor("codex");
    assert.ok(adapter instanceof CodexAdapter);
  });

  it("returns CustomPythonAdapter for custom-python: prefix", () => {
    const adapter = adapterFor("custom-python:./Agent.py");
    assert.ok(adapter instanceof CustomPythonAdapter);
    assert.deepStrictEqual(adapter._commandTemplate, ["python3", "./Agent.py", "{target_url}"]);
  });

  it("falls back to CustomCommandAdapter for unknown identifier", () => {
    const adapter = adapterFor("echo FLAG{MixedCase}");
    assert.ok(adapter instanceof CustomCommandAdapter);
    assert.deepStrictEqual(adapter._commandTemplate, ["echo", "FLAG{MixedCase}"]);
  });
});

describe("HermesAdapter", () => {
  it("sets correct command template", () => {
    const adapter = new HermesAdapter({ hermesBin: "hermes" });
    assert.ok(adapter._commandTemplate.includes("-z"));
    assert.ok(adapter._commandTemplate.includes("--yolo"));
    const templateStr = adapter._commandTemplate.join(" ");
    assert.ok(templateStr.includes("{target_url}"));
  });

  it("adds model flag when specified", () => {
    const adapter = new HermesAdapter({ hermesBin: "hermes", model: "claude-opus-4-8" });
    assert.ok(adapter._commandTemplate.includes("-m"));
    assert.ok(adapter._commandTemplate.includes("claude-opus-4-8"));
  });
});

describe("OpenClawAdapter", () => {
  it("uses infer model run command", () => {
    const adapter = new OpenClawAdapter({ openclawBin: "echo" });
    assert.deepStrictEqual(
      adapter._commandTemplate.slice(0, 4),
      ["echo", "infer", "model", "run"],
    );
    assert.ok(adapter._commandTemplate.includes("--local"));
    assert.ok(adapter._commandTemplate.includes("--json"));
  });

  it("adds model flag when specified", () => {
    const adapter = new OpenClawAdapter({ openclawBin: "echo", model: "claude-sonnet-4-6" });
    assert.ok(adapter._commandTemplate.includes("--model"));
    assert.ok(adapter._commandTemplate.includes("claude-sonnet-4-6"));
  });
});

describe("PiAdapter", () => {
  it("includes print and json mode flags", () => {
    const adapter = new PiAdapter({ piBin: "echo" });
    assert.ok(adapter._commandTemplate.includes("--print"));
    assert.ok(adapter._commandTemplate.includes("--mode"));
    assert.ok(adapter._commandTemplate.includes("json"));
    const templateStr = adapter._commandTemplate.join(" ");
    assert.ok(templateStr.includes("{target_url}"));
  });

  it("adds model flag when specified", () => {
    const adapter = new PiAdapter({ piBin: "echo", model: "claude-sonnet-4-6" });
    assert.ok(adapter._commandTemplate.includes("--model"));
    assert.ok(adapter._commandTemplate.includes("claude-sonnet-4-6"));
  });
});

describe("OpenCLIAdapter", () => {
  it("uses browser extract command", () => {
    const adapter = new OpenCLIAdapter({ opencliBin: "opencli" });
    assert.deepStrictEqual(adapter._commandTemplate, ["opencli", "browser", "extract"]);
  });
});

describe("CodexAdapter", () => {
  it("uses codex exec json command", () => {
    const adapter = new CodexAdapter({ codexBin: "codex" });
    assert.deepStrictEqual(adapter._commandTemplate.slice(0, 3), ["codex", "exec", "--json"]);
    assert.ok(adapter._commandTemplate.join(" ").includes("{target_url}"));
  });
});

describe("detectAvailableAdapters", () => {
  it("returns object with expected keys", () => {
    const available = detectAvailableAdapters();
    assert.ok("hermes" in available);
    assert.ok("openclaw" in available);
    assert.ok("opencli" in available);
    assert.ok("codex" in available);
    assert.ok("pi" in available);
  });

  it("uses argv-based lookup without shell", () => {
    const calls = [];
    const available = detectAvailableAdapters((command, args, options) => {
      calls.push({ command, args, shell: options.shell });
      return { status: args[0] === "python3" ? 0 : 1 };
    });
    assert.deepStrictEqual(available, {
      hermes: false,
      openclaw: false,
      opencli: false,
      codex: false,
      pi: false,
      python3: true,
    });
    assert.ok(calls.every((call) => call.args.length === 1));
    assert.ok(calls.every((call) => call.shell === false));
  });
});
